pkg/storage/swift/swift.go

```diff
@@ -1,84 +1,206 @@
 package swift
 
 import (
+	"fmt"
+	"math/rand"
+	"reflect"
+
 	corev1 "k8s.io/api/core/v1"
 
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack"
+	"github.com/gophercloud/gophercloud/openstack/objectstorage/v1/containers"
+	"github.com/gophercloud/utils/openstack/clientconfig"
+
+	operatorapi "github.com/openshift/api/operator/v1"
+
 	imageregistryv1 "github.com/openshift/cluster-image-registry-operator/pkg/apis/imageregistry/v1"
 	regopclient "github.com/openshift/cluster-image-registry-operator/pkg/client"
+	"github.com/openshift/cluster-image-registry-operator/pkg/clusterconfig"
+	"github.com/openshift/cluster-image-registry-operator/pkg/storage/util"
 )
 
 type driver struct {
-	Config  *imageregistryv1.ImageRegistryConfigStorageSwift
+	// Config is a struct where the basic configuration is stored
+	Config *imageregistryv1.ImageRegistryConfigStorageSwift
+	// Listers are used to download OpenStack credentials from the native secret
 	Listers *regopclient.Listers
 }
 
+// getSwiftClient returns a client that allows to interact with the OpenStack Swift service
+func (d *driver) getSwiftClient(cr *imageregistryv1.Config) (*gophercloud.ServiceClient, error) {
+	clientOpts := new(clientconfig.ClientOpts)
+	var opts *gophercloud.AuthOptions
+	cfg, err := clusterconfig.GetSwiftConfig(d.Listers)
+	if err != nil {
+		return nil, err
+	}
+
+	// In a case of IPI all credentials are stored in clouds.yaml file in the secret, prepared by
+	// cloud-credentials-operator, and we read them from cfg.Storage.Swift.Cloud.
+	// For UPI it's enough to provide username and password in the related secret.
+	if cfg.Storage.Swift.Cloud != nil {
+		cloud := cfg.Storage.Swift.Cloud
+
+		if cloud.AuthInfo != nil {
+			clientOpts.AuthInfo = cloud.AuthInfo
+			clientOpts.AuthType = cloud.AuthType
+			clientOpts.Cloud = cloud.Cloud
+			clientOpts.RegionName = cloud.RegionName
+		}
+
+		opts, err = clientconfig.AuthOptions(clientOpts)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		opts = &gophercloud.AuthOptions{
+			IdentityEndpoint: cr.Spec.Storage.Swift.AuthURL,
+			Username:         cfg.Storage.Swift.Username,
+			Password:         cfg.Storage.Swift.Password,
+			DomainID:         cr.Spec.Storage.Swift.DomainID,
+			DomainName:       cr.Spec.Storage.Swift.Domain,
+			TenantID:         cr.Spec.Storage.Swift.TenantID,
+			TenantName:       cr.Spec.Storage.Swift.Tenant,
```

Comment:
@adambkaplan you were advocating for keeping this options and not to replacing them with cloud.yaml, and I don't know your vision of their interaction when you have cloud.yaml and some fields in the custom resource, so please review this.